daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/svalue/Ordering.scala

```diff
@@ -5,22 +5,105 @@ package com.daml.lf
 package speedy
 package svalue
 
-import com.daml.lf.data.{Bytes, FrontStack, FrontStackCons, ImmArray, Utf8}
-import com.daml.lf.data.ScalazEqual._
-import com.daml.lf.language.TypeOrdering
-import com.daml.lf.speedy.SError.SErrorCrash
-import com.daml.lf.speedy.SValue._
-import com.daml.lf.value.Value.ContractId
+import data.{Bytes, Utf8}
+import language.TypeOrdering
+import value.Value.ContractId
 
-import scala.annotation.tailrec
 import scala.collection.JavaConverters._
 
 object Ordering extends scala.math.Ordering[SValue] {
 
-  private def compareText(text1: String, text2: String): Int =
-    Utf8.Ordering.compare(text1, text2)
+  @throws[SError.SErrorCrash]
+  // Ordering between two SValues of same type.
+  // This follows the equality defined in the daml-lf spec.
+  def compare(x: SValue, y: SValue): Int = {
+    import SValue._
 
-  private def compareCid(cid1: ContractId, cid2: ContractId): Int =
+    var diff = 0
+    var stackX = List(Iterator.single(x))
+    var stackY = List(Iterator.single(y))
+    // invariant: stackX.length == stackY.length
+
+    @inline
+    def push(xs: Iterator[SValue], ys: Iterator[SValue]): Unit = {
+      stackX = xs :: stackX
+      stackY = ys :: stackY
+    }
+
+    @inline
+    def pop(): Unit = {
+      stackX = stackX.tail
+      stackY = stackY.tail
+    }
+
+    @inline
+    def step(tuple: (SValue, SValue)) =
+      tuple match {
+        case (SUnit, SUnit) =>
```

Comment:
I know this is not necessary but it makes the code easier to read for people with limited Scala experience, of which we have many.
```suggestion
        case (SUnit, SUnit) =>
          ()
```